### ISSUE TITLE
DecoupledWeightDecayExtension only decays weights which are present in the sparse gradient

### ISSUE THREAD

**Issue by bmc2-stripe** [tensorflow/addons#767]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 16.04
- TensorFlow version and how it was installed (source or binary): 2.0, binary
- TensorFlow-Addons version and how it was installed (source or binary): 0.6.0, binary
- Python version: 3.7
- Is GPU used? (yes/no): yes

**Describe the bug**
The current implementation only makes updates to those indices whose present in the gradient's IndexedSlices:

https://github.com/tensorflow/addons/blob/v0.6.0/tensorflow_addons/optimizers/weight_decay_optimizers.py#L164

In the paper which motivated this approach, no mention is made of special handling for sparse vectors (https://arxiv.org/pdf/1711.05101.pdf). Indeed, if the spirit of the paper is to apply weight decay in manner decoupled from gradient-based updates, it seems odd that we would zero-out decay whenever a loss gradient is not present.

A simple fix would be to apply _decay_weights_op to both sparse and dense gradients. I am happy to send a PR if the maintainers are open to it.

**Comment by seanpmorgan** [comment]
cc @PhilJd codeowner.

**Comment by PhilJd** [comment]
Hi @bmc2-stripe,
I never personally used sparse gradients so I don't know if sparse or dense decay performs better but there was some discussion in the initial PR here: https://github.com/tensorflow/tensorflow/pull/17438#discussion_r182403534. The main motivation to go for this was that LazyAdam, which updates the accumulators only for the indices present works better for sparse gradients
Do you have experiments where the dense update for sparse gradients works better? If so, I think it would be reasonable to add an init parameter `force_dense_update`, which if set to true implements the behaviour you describe.

**Comment by bmc2-stripe** [comment]
Thank you both for taking a look, and the the pointer to the review thread! An init param (set by default to False, for back-compat) would definitely help with our use case.

As for an experiment: I ran dense weight decay versus sparse weight decay on a dataset of fraudulent payment activity we have at Stripe. Our primary model performance metric is partial AUC. On that metric, the dense implementation outperforms by +7% (relative), which is pretty dramatic. One caveat: this fraud model was tuned using a weight decay implementation with dense updates, so its quite possible that if we retuned it on top of sparse updates, we could recover at least some of that gap.  I'm afraid I don't have any experiment results on publicly available datasets.

**Comment by PhilJd** [comment]
If that's the case I'd be happy to merge a PR for this behavior :)

**Comment by bmc2-stripe** [comment]
Apologies for the delay on my end for this one. I'm aiming to get this sent this week, or next week at the latest.

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
